File: tools/NL/networkLauncher.js
const exec = require('child_process').execSync;
const fs = require('fs');
const path = require('path');
const winston = require('winston');
const yaml = require('js-yaml')
let YAML = require('json2yaml')
    , ymlText
    ;
/*const logger = new (winston.Logger)({
    transports: [
        new winston.transports.Console({
            level: 'debug',
            handleExceptions: true,
            prettyPrint: true,
            colorize: true
        })
    ],
    exitOnError: false
});*/
let argv = require('yargs')
    .options({
        mode: {
            demand: false,
            description: 'Mode to be used and supported options are up, down, restart, addorg, addpeer',
            alias: 'm',
            default: 'up',
            type: 'string'
        },
        input: {
            demand: true,
            description: 'Network spec input file path',
            alias: 'i',
            type: 'string'
        },
        kubeconfig: {
            demand: false,
            description: 'Kube config file path',
            alias: 'k',
            type: 'string'
        }
    })
    .help()
    .argv;

let networkSpec = yaml.safeLoad(fs.readFileSync(argv.input, 'utf8'));
async function networkLauncher(argv) {
    try {
        if (argv.mode == "up") {
            await genCrypto(networkSpec);
            await genConfigtx(networkSpec);
            await genNetwork(networkSpec);
            await genConnProfile(networkSpec);
        } else if (argv.mode == "down") {
            //TODO
        } else if (argv.mode == "restart") {
            //TODO
        } else if (argv.mode == "addpeer") {
            //TODO
        }
    } catch (err) {
        //TODO
    }
}

async function genCrypto(networkSpec) {
    try {
        let cryptoConfig = {}
        let ordSpecs = []
        let peerSpecs = []
        let OrdererOrgs = []
        let PeerOrgs = []

        for (var i = 0; i < networkSpec.orderer_organizations.length; i++) {
            ordSpecs = []
            for (var j = 0; j < networkSpec.orderer_organizations[i].num_orderers; j++) {
                ordSpecs.push({
                    "Hostname": "orderer" + j + "-" + networkSpec.orderer_organizations[i].name
                })
            }
            OrdererOrgs.push({
                "Name": networkSpec.orderer_organizations[i].name,
                "Domain": networkSpec.orderer_organizations[i].name,
                "EnableNodeOUs": true,
                "Specs": ordSpecs
            })
        }

        for (var k = 0; k < networkSpec.peer_organizations.length; k++) {
            peerSpecs = []
            for (var l = 0; l < networkSpec.peer_organizations[k].num_peers; l++) {
                peerSpecs.push({
                    "Hostname": "peer" + l + "-" + networkSpec.peer_organizations[k].name
                })
            }
            PeerOrgs.push({
                "Name": networkSpec.peer_organizations[k].name,
                "Domain": networkSpec.peer_organizations[k].name,
                "EnableNodeOUs": true,
                "Specs": peerSpecs
            })
        }

        cryptoConfig = {
            "OrdererOrgs": OrdererOrgs,
            "PeerOrgs": PeerOrgs
        }

        ymlText = YAML.stringify(cryptoConfig);
        fs.writeFileSync('./crypto-config.yaml', ymlText)
        await exec('cryptogen generate --config=./crypto-config.yaml --output=' + networkSpec.certs_location + '/crypto-config')
    } catch (err) {
        //TODO
    }
}

async function genConfigtx(networkSpec) {

    let Organizations = []
    let ordererOrganizations = []
    let peerOrganizations = []
    let OrdererAddresses = []
    let Consenters = []
    try {
        for (let i = 0; i < networkSpec.orderer_organizations.length; i++) {
            let orderer_organization = networkSpec.orderer_organizations[i]
            for (let j = 0; j < orderer_organization.num_orderers; j++) {
                i = i > 0 ? i : ""
                OrdererAddresses.push("orderer" + j + "-" + orderer_organization.name + ":7050")
                if (networkSpec.orderer.orderertype == "etcdraft") {
                    Consenters.push({
                        "Host": "orderer" + j + "-" + orderer_organization.name,
                        "Port": 7050,
                        "ClientTLSCert": networkSpec.certs_location + "crypto-config/ordererOrganizations/" + orderer_organization.name + "/orderers/orderer" + j + "-" + orderer_organization.name + "." + orderer_organization.name + "/tls/server.crt",
                        "ServerTLSCert": networkSpec.certs_location + "crypto-config/ordererOrganizations/" + orderer_organization.name + "/orderers/orderer" + j + "-" + orderer_organization.name + "." + orderer_organization.name + "/tls/server.crt",
                    })
                }
            }
            ordererOrganizations.push(addOrganizationsForConfigtx(orderer_organization, "orderer", networkSpec.certs_location))
        }
        for (let i = 0; i < networkSpec.peer_organizations.length; i++) {
            let peerer_organization = networkSpec.peer_organizations[i]
            peerOrganizations.push(addOrganizationsForConfigtx(peerer_organization, "peer", networkSpec.certs_location))
        }
        Organizations = ordererOrganizations.concat(peerOrganizations)
        let capabilities = {
            "Global": {
                "V1_3": true
            },
            "Orderer": {
                "V1_1": true
            },
            "Application": {
                "V1_3": true
            }
        }

        let Orderer = {
            "OrdererType": networkSpec.orderer.orderertype,
            "Addresses": OrdererAddresses,
            "BatchTimeout": networkSpec.orderer.batchtimeout,
            "BatchSize": {
                "MaxMessageCount": networkSpec.orderer.batchsize.maxmessagecount,
                "AbsoluteMaxBytes": networkSpec.orderer.batchsize.absolutemaxbytes,
                "PreferredMaxBytes": networkSpec.orderer.batchsize.preferredmaxbytes
            },
            "Organizations": ordererOrganizations,
            "Policies": {
                "Readers": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Readers"
                },
                "Writers": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Writers"
                },
                "Admins": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Admins"
                },
                "BlockValidation": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Writers"
                }
            },
            "Capabilities": {
                "V1_1": true
            }
        }
        let EtcdRaft = {
            "Consenters": Consenters,
            "Options": {
                "TickInterval": networkSpec.orderer.etcdraft_options.TickInterval,
                "ElectionTick": networkSpec.orderer.etcdraft_options.ElectionTick,
                "HeartbeatTick": networkSpec.orderer.etcdraft_options.HeartbeatTick,
                "MaxInflightBlocks": networkSpec.orderer.etcdraft_options.MaxInflightBlocks,
                "SnapshotIntervalSize": networkSpec.orderer.etcdraft_options.SnapshotIntervalSize
            }
        }
        if (networkSpec.orderer.orderertype == "etcdraft") {
            Orderer["EtcdRaft"] = EtcdRaft
        }

        let Application = {
            "Organizations": peerOrganizations,
            "Policies": {
                "Readers": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Readers"
                },
                "Writers": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Writers"
                },
                "Admins": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Admins"
                }
            },
            "Capabilities": {
                "V1_3": true
            }
        }

        let Channel = {
            "Policies": {
                "Readers": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Readers"
                },
                "Writers": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Writers"
                },
                "Admins": {
                    "Type": "ImplicitMeta",
                    "Rule": "ANY Admins"
                }
            },
            "Capabilities": {
                "V1_3": true
            }
        }

        let Profiles = {
            "testorgschannel": {
                "Policies": Channel["Policies"],
                "Capabilities": Channel["Capabilities"],
                "Consortium": "FabricConsortium",
                "Application": {
                    "Organizations": peerOrganizations,
                    "Policies": Channel["Policies"],
                    "Capabilities": Channel["Capabilities"]
                },
                "Orderer": Orderer
            },
            "testOrgsOrdererGenesis": {
                "Policies": Channel["Policies"],
                "Capabilities": Channel["Capabilities"],
                "Orderer": Orderer,
                "Consortiums": {
                    "FabricConsortium": {
                        "Organizations": peerOrganizations
                    }
                }
            }
        }

        const configtxJson = {
            "Organizations": Organizations,
            "Capabilities": capabilities,
            "Orderer": Orderer,
            "Application": Application,
            "Channel": Channel,
            "Profiles": Profiles
        }
        ymlText = YAML.stringify(configtxJson);
        fs.writeFileSync('./configtx.yaml', ymlText)

        await exec('configtxgen -profile testOrgsOrdererGenesis -channelID ordersystemchannel -outputBlock ' + networkSpec.certs_location + 'crypto-config/ordererOrganizations/genesis.block')
        for (let i = 0; i < networkSpec.num_channels; i++) {
            await exec('configtxgen -profile testorgschannel -channelCreateTxBaseProfile testOrgsOrdererGenesis -channelID testorgschannel'+i+' -outputCreateChannelTx ./testorgschannel1.tx')
        }
    } catch (err) {
        //TODO
    }
}

async function genNetwork(networkSpec) {
    try {
        //TODO
    } catch (err) {
        //TODO
    }
}

async function genConnProf(networkSpec) {
    try {
        //TODO
    } catch (err) {
        //TODO
    }
}

function addOrganizationsForConfigtx(organization, type, certs_location) {
    let peerPolicies = {
        "Readers": {
            "Type": "Signature",
            "Rule": "OR('" + organization.name + ".admin', '" + organization.name + ".peer')"
        },
        "Writers": {
            "Type": "Signature",
            "Rule": "OR('" + organization.name + ".admin', '" + organization.name + ".client')"
        },
        "Admins": {
            "Type": "Signature",
            "Rule": "OR('" + organization.name + ".admin')"
        }
    }
    let ordererPolicies = {
        "Readers": {
            "Type": "Signature",
            "Rule": "OR('" + organization.name + ".member')"
        },
        "Writers": {
            "Type": "Signature",
            "Rule": "OR('" + organization.name + ".member')"
        },
        "Admins": {
            "Type": "Signature",
            "Rule": "OR('" + organization.name + ".admin')"
        }
    }
    let Policies = type == "peer" ? peerPolicies : ordererPolicies
    let organizationData = {
        "Name": organization.name,
        "ID": organization.name,
        "MSPDir": certs_location + "crypto-config/" + type + "Organizations/" + organization.name + "/msp",
        "Policies": Policies
    }
    return organizationData
}

networkLauncher(argv);

File: tools/NL/go/networkLauncher.go
package main

import (
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"text/template"

	"github.com/hyperledger/fabric-test/tools/NL/go/templates"
	yaml "gopkg.in/yaml.v2"
)

type NetworkLauncher struct {
}

// Config Struct
type Config struct {
	DbType               string                 `yaml:"db_type,omitempty"`
	PeerLoggingSPec      string                 `yaml:"peer_fabric_logging_spec,omitempty"`
	OrdererLoggingSPec   string                 `yaml:"orderer_fabric_logging_spec,omitempty"`
	TLS                  string                 `yaml:"tls,omitempty"`
	Metrics              bool                   `yaml:"metrics,omitempty"`
	ArtifactsLocation    string                 `yaml:"certs_location,omitempty"`
	OrdererConfig        OrdererConfig          `yaml:"orderer,omitempty"`
	KafkaConfig          KafkaConfig            `yaml:"kafka,omitempty"`
	OrdererOrganizations []OrdererOrganizations `yaml:"orderer_organizations,omitempty"`
	PeerOrganizations    []PeerOrganizations    `yaml:"peer_organizations,omitempty"`
	NumChannels          int                    `yaml:"num_channels,omitempty"`
}

type ConfigTx struct {
	OrdererOrganizations []OrdererOrganizations `yaml:"orderer_organizations,omitempty"`
	PeerOrganizations    []PeerOrganizations    `yaml:"peer_organizations,omitempty"`
	OrdererAddresses     []string               `yaml:"orderer_addresses,omitempty"`
	OrdererConfig        OrdererConfig          `yaml:"orderer,omitempty"`
	OrdererCerts         []OrdererCert          `yaml:"orderer_certs,omitempty"`
	ArtifactsLocation    string                 `yaml:"certs_location,omitempty"`
}

type OrdererCert struct {
	Host            string `yaml:"host,omitempty"`
	TlsCertLocation string `yaml:"tls_cert_location,omitempty"`
}

// OrdererConfig struct
type OrdererConfig struct {
	OrdererType string
	Batchsize   struct {
		Maxmessagecount   int    `yaml:"maxmessagecount,omitempty"`
		Absolutemaxbytes  string `yaml:"absolutemaxbytes,omitempty"`
		Preferredmaxbytes string `yaml:"preferredmaxbytes,omitempty"`
	} `yaml:"batchsize,omitempty"`
	Batchtimeout    string
	Etcdraftoptions struct {
		TickInterval         string `yaml:"TickInterval,omitempty"`
		ElectionTick         int    `yaml:"ElectionTick,omitempty"`
		HeartbeatTick        int    `yaml:"HeartbeatTick,omitempty"`
		MaxInflightBlocks    int    `yaml:"MaxInflightBlocks,omitempty"`
		SnapshotIntervalSize string `yaml:"SnapshotIntervalSize,omitempty"`
	} `yaml:"etcdraft_options,omitempty"`
}

// OrdererOrganizations struct
type OrdererOrganizations struct {
	Name        string `yaml:"name,omitempty"`
	MspId       string `yaml:"msp_id,omitempty"`
	NumOrderers int    `yaml:"num_orderers,omitempty"`
	NumCa       int    `yaml:"num_ca,omitempty"`
}

// KafkaConfig struct
type KafkaConfig struct {
	NumKafka             int `yaml:"num_kafka,omitempty"`
	NumKafkaReplications int `yaml:"num_kafka_replications,omitempty"`
	NumZookeepers        int `yaml:"num_zookeepers,omitempty"`
}

// PeerOrganizations stuct
type PeerOrganizations struct {
	Name     string `yaml:"name,omitempty"`
	MspId    string `yaml:"msp_id,omitempty"`
	NumPeers int    `yaml:"num_peers,omitempty"`
	NumCa    int    `yaml:"num_ca,omitempty"`
}

func (n *NetworkLauncher) getConf(networkSpecPath string) Config {
	var config Config
	yamlFile, err := ioutil.ReadFile(networkSpecPath)
	if err != nil {
		log.Fatalf("Failed to read input file %v", err)
	}
	err = yaml.Unmarshal(yamlFile, &config)
	if err != nil {
		log.Fatalf("Failed to create config object %v", err)
	}
	return config
}

func (n *NetworkLauncher) getOrdererAddresses(networkSpec Config) ([]string, []OrdererCert) {
	var orderers []string
	var ordererCerts []OrdererCert
	for i := 0; i < len(networkSpec.OrdererOrganizations); i++ {
		for j := 0; j < networkSpec.OrdererOrganizations[i].NumOrderers; j++ {
			orderers = append(orderers, fmt.Sprintf("orderer%v-%s:7050", j, networkSpec.OrdererOrganizations[i].Name))
			if networkSpec.OrdererConfig.OrdererType == "etcdraft" {
				hostName := fmt.Sprintf("orderer%v-%s", j, networkSpec.OrdererOrganizations[i].Name)
				tlsCertLocation := fmt.Sprintf("%vcrypto-config/ordererOrganizations/%v/orderers/%v.%v/tls/server.crt", networkSpec.ArtifactsLocation, networkSpec.OrdererOrganizations[i].Name, hostName, networkSpec.OrdererOrganizations[i].Name)
				ordererCert := OrdererCert{Host: hostName, TlsCertLocation: tlsCertLocation}
				ordererCerts = append(ordererCerts, ordererCert)
			}
		}
	}
	return orderers, ordererCerts
}

func (n *NetworkLauncher) Iterator(num int) []int{
	var result []int
	for i := 0; i < num; i++ {
		result = append(result,i)
	}
	return result
}

func (n *NetworkLauncher) GenerateConfigtx(networkSpec Config) error {

	var NL NetworkLauncher
	ordererAddresses, ordererCerts := NL.getOrdererAddresses(networkSpec)

	configTxInput := ConfigTx{OrdererOrganizations: networkSpec.OrdererOrganizations, PeerOrganizations: networkSpec.PeerOrganizations, OrdererAddresses: ordererAddresses, OrdererConfig: networkSpec.OrdererConfig, OrdererCerts: ordererCerts, ArtifactsLocation: networkSpec.ArtifactsLocation}
	err := NL.GenerateConfigFiles(configTxInput)

	if err != nil {
		return fmt.Errorf("%v", err)
	}
	return nil
}

func (n *NetworkLauncher) GenerateConfigFiles(input interface{}) error {

	var NL NetworkLauncher
	var fileName string
	var action string
	var fileTemplate string

	switch fmt.Sprintf("%T", input) {
	case "main.ConfigTx":
		fileName = "configtx.yaml"
		action = "configtx"
		fileTemplate = templates.DefaultConfigTxTemplate()
	}

	currentDir, err := os.Getwd()
	if err != nil {
		log.Fatal(err)
	}

	config, err := os.Create(filepath.Join(currentDir, "./configFiles/", fileName))
	if err != nil {
		return fmt.Errorf("Failed to create %v path", action)
	}

	defer config.Close()
	var templateResult *template.Template
	templateResult, err = template.New(action).Funcs(template.FuncMap{"N": NL.Iterator}).Parse(fileTemplate)

	if err != nil {
		return fmt.Errorf("Failed to parse %v template", action)
	}

	err = templateResult.Execute(io.MultiWriter(config), input)
	if err != nil {
		return fmt.Errorf("Failed to create %v.yaml", action)
	}

	return nil
}


func (n *NetworkLauncher) GenerateConfigurationFiles() error {

	cmd := exec.Command("./ytt-linux-amd64 ", "-f", "./templates/", "--output", "./configFiles")

	stdoutStderr, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("%v", err)
	}
	fmt.Printf(string(stdoutStderr))
	return nil
}

func (n *NetworkLauncher) GenerateCryptoCerts(networkSpec Config) error {

	cmd := exec.Command("cryptogen", "generate", "--config=./configFiles/crypto-config.yaml", fmt.Sprintf("--output=%vcrypto-config", networkSpec.ArtifactsLocation))

	stdoutStderr, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("%v", err)
	}
	fmt.Printf(string(stdoutStderr))
	return nil
}

func (n *NetworkLauncher) GenerateOrdererGenesisBlock() error {
	cmd := exec.Command("configtxgen", "-profile", "testOrgsOrdererGenesis", "-channelID", "ordersystemchannel", "-outputBlock", "./genesis.block")
	stdoutStderr, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("%v", err)
	}
	fmt.Printf(string(stdoutStderr))
	return nil
}

func (n *NetworkLauncher) GenerateChannelTransaction(networkSpec Config) error {

	for i := 0; i < networkSpec.NumChannels; i++ {
		cmd := exec.Command("configtxgen", "-profile", "testorgschannel", "-channelCreateTxBaseProfile", "testOrgsOrdererGenesis", "-channelID", fmt.Sprintf("testorgschannel%v", i), "-outputCreateChannelTx", fmt.Sprintf("./testorgschannel%v.tx", i))
		stdoutStderr, err := cmd.CombinedOutput()
		if err != nil {
			return fmt.Errorf("%v", err)
		}
		fmt.Printf(string(stdoutStderr))
	}
	return nil
}

func (n *NetworkLauncher) ReadArguments() (*string, *string) {

	networkSpecPath := flag.String("i", "", "Network spec input file path")
	kubeConfigPath := flag.String("k", "", "Kube config file path")

	flag.Parse()
	if fmt.Sprintf("%s", *kubeConfigPath) == "" {
		fmt.Println("Kube config file path not provided")
	}
	if fmt.Sprintf("%s", *networkSpecPath) == "" {
		log.Fatalf("Network spec file path not provided")
	}

	return networkSpecPath, kubeConfigPath
}

func main() {
	var NL NetworkLauncher

	networkSpecPath, _ := NL.ReadArguments()
	input := NL.getConf(*networkSpecPath)

	err := NL.GenerateConfigurationFiles()
	if err != nil {
		log.Fatalf("Failed to generate yaml files%v", err)
	}

	err = NL.GenerateCryptoCerts(input)
	if err != nil {
		log.Fatalf("Failed to generate certificates %v", err)
	}

	err = NL.GenerateConfigtx(input)
	if err != nil {
		log.Fatalf("Failed to create configtx.yaml file %v", err)
	}

	err = NL.GenerateOrdererGenesisBlock()
	if err != nil {
		log.Fatalf("Failed to create orderer genesis block %v", err)
	}
	err = NL.GenerateChannelTransaction(input)
	if err != nil {
		log.Fatalf("Failed to create channel transaction %v", err)
	}
}
